infrastructure/ansible/dynamic.inventory/TO.py

```diff
@@ -0,0 +1,241 @@
+                    server['cachegroupId'])
+                out[flat_cachegroup] = cgdata.cgvars
+                flat_parent_cg = cgdata.primary_parent_group_name
+                flat_second_parent_cg = cgdata.secondary_parent_group_name
+                if flat_parent_cg not in out.keys():
+                    out[flat_parent_cg] = {}
+                    out[flat_parent_cg]['children'] = []
+                if flat_second_parent_cg not in out.keys():
+                    out[flat_second_parent_cg] = {}
+                    out[flat_second_parent_cg]['children'] = []
+                out[flat_parent_cg]['children'].append(flat_cachegroup)
+                out[flat_second_parent_cg]['children'].append(flat_cachegroup)
+            out[flat_cachegroup]['hosts'].append(fqdn)
+            if flat_server_type not in out.keys():
+                out['server_type']['children'].append(flat_server_type)
+                out[flat_server_type] = {}
+                out[flat_server_type]['hosts'] = []
+            out[flat_server_type]['hosts'].append(fqdn)
+            if flat_server_cdn_name not in out.keys():
+                out['server_cdnName']['children'].append(flat_server_cdn_name)
+                out[flat_server_cdn_name] = {}
+                out[flat_server_cdn_name]['hosts'] = []
+            out[flat_server_cdn_name]['hosts'].append(fqdn)
+            if flat_server_status not in out.keys():
+                out['server_status']['children'].append(flat_server_status)
+                out[flat_server_status] = {}
+                out[flat_server_status]['hosts'] = []
+            out[flat_server_status]['hosts'].append(fqdn)
+        return out
+
+    def to_inventory(self):
+        return self.generate_inventory_list(self.to_url)
+
+
+if __name__ == "__main__":
+    PARSER = argparse.ArgumentParser(
+        description='Generate ansible inventory from TrafficOps')
+
+    PARSER.add_argument(
+        '-to',
+        nargs='+',
+        help='Please pass username and password for TrafficOps')
+    PARSER.add_argument(
+        '-url',
+        type=str,
+        help='TO URL')
+    PARSER.add_argument('--list', action='store_true')
+    PARSER.add_argument('--host', type=str, help='Target TO Server')
+
+    ARGS = PARSER.parse_args()
+
+    TO_USER = os.getenv("TO_USERNAME")
+    TO_PASS = os.getenv("TO_PASSWORD")
+    TO_URL = os.getenv("TO_URL")
+
+    if ARGS.to:
+        if TO_USER is None:
+            TO_USER = ARGS.to[0]
+
+        if TO_PASS is None:
+            TO_PASS = ARGS.to[1]
+
+    if TO_URL is None:
+        if ARGS.env is None:
+            TO_URL = "https://localhost:8080"
+        else:
+            TO_URL = ARGS.env
+
+    if TO_USER and TO_PASS is not None:
+        if ARGS.list:
+            TMP_ANSIBLEINVENTORY = AnsibleInventory(TO_USER, TO_PASS, TO_URL)
+            INVENTORY = TMP_ANSIBLEINVENTORY.to_inventory()
+        # Since we're supplying hostvar metadata, --host support isn't required
```

Comment:
Should you remove the flag, then?

Reply:
It's required as part of the calling convention.

https://docs.ansible.com/ansible/2.6/dev_guide/developing_inventory.html